### PR TITLE
Adding adjoint methods to FermiWord and FermiSentence classes

### PR DESCRIPTION
**Pull request by willjmax**
This PR adds a `.adjoint()` method to the `FermiWord` and `FermiSentence` classes. In both cases the method can be used in the following way: `f_dag = f.adjoint()`. The resulting `f_dag` is a `FermiWord` or `FermiSentence` corresponding to the adjoint of `f`. This PR addresses [sc-72055].

### CODE REVIEW

File: pennylane/fermi/fermionic.py [hunk 1]
```diff
@@ -55,6 +55,18 @@ def __init__(self, operator):
...
+        r"""Return the adjiont of the FermiWord"""
+        n = len(self.items())
+        adjoint_dict = {}
+        for key, value in self.items():
+            position = n - key[0] - 1
+            orbital = key[1]
+            fermi = "+" if value == "-" else "-"
+            adjoint_dict[(position, orbital)] = fermi
+
+        return FermiWord(adjoint_dict)
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:68]: This looks okay, but I have one concern regarding the dictionary object we built here. The order of the items in the conjugated Fermi word is opposite to what we usually expect.
image

File: pennylane/fermi/fermionic.py [hunk 2]
```diff
@@ -326,6 +338,17 @@ class FermiSentence(dict):
     def __init__(self, operator):
         super().__init__(operator)
 
+    def adjoint(self):
+        r"""Return the adjoint of the FermiSentence"""
+        adjoint_dict = {}
+        for key, value in self.items():
+            word = key.adjoint()
+            scalar = qml.math.conj(value)
+
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:347]:
```suggestion
```

**Review by obliviateandsurrender — COMMENTED** [review]
Thanks @willjmax, leaving some suggestions.

File: tests/fermi/test_fermionic.py [hunk 3]
```diff
@@ -141,6 +161,22 @@ def test_to_mat_error(self):
...
+        (fw5, fw5_dag),
+        (fw6, fw6_dag),
+        (fw7, fw7_dag),
+    )
+
+    @pytest.mark.parametrize("fw, fw_dag", tup_fw_dag)
+    def test_adjoint(self, fw, fw_dag):
+        assert fw.adjoint() == fw_dag
+        assert FermiA(0).adjoint() == FermiC(0)
+        assert FermiC(0).adjoint() == FermiA(0)
```
Comment (austingmhuang) [inline tests/fermi/test_fermionic.py:178]: Do we need this? Seems like it's constant across all the parameters. Could add this as another test case or just delete it altogether.

Comment (willjmax) [reply tests/fermi/test_fermionic.py:178]: It doesn't seem necessary, but I added it because the code coverage report complained about these methods not having tests.

**Comment by willjmax** [comment]
> Looking good @willjmax.
> 
> Do we need to add test cases for `FermiWord` specifically? I see the tests for FermiSentence but it might be useful to have unit tests for `adjoint()` of `FermiWord` as well.

There are tests for both `FermiWord` and `FermiSentence`.

File: pennylane/fermi/fermionic.py (hunk 1 above)
Comment (willjmax) [reply pennylane/fermi/fermionic.py:68]: `.items()` will show the order used to build the `FermiWord`, but `FermiWord` has a `.sorted_dict` property which displays the proper order. Maybe we could change `.items()` to return `self.sorted_dic.items()` instead?

Comment (obliviateandsurrender) [reply pennylane/fermi/fermionic.py:68]: or we could iterate over `reversed(self.items())` here 🤔

Comment (austingmhuang) [reply pennylane/fermi/fermionic.py:68]: Iterating backwards seems easier to me.

File: tests/fermi/test_fermionic.py (hunk 3 above)
Comment (austingmhuang) [reply tests/fermi/test_fermionic.py]: Ah I see. I would just write this out as a separate test e.g. test_adjoint_fermiac instead.

Comment (austingmhuang) [reply tests/fermi/test_fermionic.py]: Oh I saw your changes, that works as well.

**Comment by austingmhuang** [comment]
> > Looking good @willjmax.
> > Do we need to add test cases for `FermiWord` specifically? I see the tests for FermiSentence but it might be useful to have unit tests for `adjoint()` of `FermiWord` as well.
> 
> There are tests for both `FermiWord` and `FermiSentence`.

My bad, think I scrolled too quickly 😅. Looks good to me for the most part. Could add docstrings for the new tests but they seem straightforward enough to me that I personally wouldn't require them. @obliviateandsurrender are docstrings a hard requirement or is it ok to omit them for really simple tests?

**Review by austingmhuang — APPROVED** [review]
Nice work @willjmax

File: pennylane/fermi/fermionic.py [hunk 4]
```diff
@@ -326,6 +341,16 @@ class FermiSentence(dict):
     def __init__(self, operator):
         super().__init__(operator)
 
+    def adjoint(self):
+        r"""Return the adjoint of the FermiSentence"""
```
Comment (austingmhuang) [inline pennylane/fermi/fermionic.py:345]:
```suggestion
        r"""Return the adjoint of the ``FermiSentence``"""
```

File: pennylane/fermi/fermionic.py [hunk 5]
```diff
@@ -647,9 +672,13 @@ def __init__(self, orbital):
             raise ValueError(
                 f"FermiC: expected a single, positive integer value for orbital, but received {orbital}"
             )
+        self.orbital = orbital
         operator = {(0, orbital): "+"}
         super().__init__(operator)
 
+    def adjoint(self):
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:679]: minor nitpick but docstring is missing here 😅

File: pennylane/fermi/fermionic.py [hunk 6]
```diff
@@ -684,5 +713,9 @@ def __init__(self, orbital):
             raise ValueError(
                 f"FermiA: expected a single, positive integer value for orbital, but received {orbital}"
             )
+        self.orbital = orbital
         operator = {(0, orbital): "-"}
         super().__init__(operator)
+
+    def adjoint(self):
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:720]: same as above

File: tests/fermi/test_fermionic.py [hunk 7]
```diff
@@ -141,6 +161,22 @@ def test_to_mat_error(self):
...
+    tup_fw_dag = (
+        (fw1, fw1_dag),
+        (fw2, fw2_dag),
+        (fw3, fw3_dag),
+        (fw4, fw4_dag),
+        (fw5, fw5_dag),
+        (fw6, fw6_dag),
+        (fw7, fw7_dag),
+        (FermiA(0), FermiC(0)),
+        (FermiC(0), FermiA(0)),
```
Comment (obliviateandsurrender) [inline tests/fermi/test_fermionic.py:173]: Can we add a test case with a non-zero orbital?

**Review by obliviateandsurrender — APPROVED** [review]
Looks good, just some minor suggestions.

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -13,13 +13,17 @@
...
+  [(#6166)](https://github.com/PennyLaneAI/pennylane/pull/6166)
+
 * Improve `qml.Qubitization` decomposition.
   [(#6182)](https://github.com/PennyLaneAI/pennylane/pull/6182)
 
 * The `__repr__` methods for `FermiWord` and `FermiSentence` now returns a
   unique representation of the object.
   [(#6167)](https://github.com/PennyLaneAI/pennylane/pull/6167)
 
+
```
Comment (obliviateandsurrender) [inline doc/releases/changelog-dev.md:26]:
```suggestion
```

File: pennylane/fermi/fermionic.py [hunk 9]
```diff
@@ -647,9 +672,14 @@ def __init__(self, orbital):
             raise ValueError(
                 f"FermiC: expected a single, positive integer value for orbital, but received {orbital}"
             )
+        self.orbital = orbital
         operator = {(0, orbital): "+"}
         super().__init__(operator)
 
+    def adjoint(self):
+        """Return the adjoint of FermiC"""
```
Comment (soranjh) [inline pennylane/fermi/fermionic.py:680]:
```suggestion
        """Return the adjoint of FermiC."""
```

File: pennylane/fermi/fermionic.py [hunk 10]
```diff
@@ -684,5 +714,10 @@ def __init__(self, orbital):
             raise ValueError(
                 f"FermiA: expected a single, positive integer value for orbital, but received {orbital}"
             )
+        self.orbital = orbital
         operator = {(0, orbital): "-"}
         super().__init__(operator)
+
+    def adjoint(self):
+        """Return the adjoint of FermiA"""
```
Comment (soranjh) [inline pennylane/fermi/fermionic.py:722]:
```suggestion
        """Return the adjoint of FermiA."""
```

File: pennylane/fermi/fermionic.py [hunk 11]
```diff
@@ -55,6 +55,21 @@ def __init__(self, operator):
...
+        adjoint_dict = {}
+        for key, value in reversed(self.items()):
+            position = n - key[0] - 1
+            orbital = key[1]
+            fermi = "+" if value == "-" else "-"
+            adjoint_dict[(position, orbital)] = fermi
+
+        return FermiWord(adjoint_dict)
+
+    def items(self):
```
Comment (soranjh) [inline pennylane/fermi/fermionic.py:70]: Please provide a doctsring.
